src/java.base/share/classes/jdk/internal/foreign/abi/AbstractLinker.java

```diff
@@ -189,54 +188,104 @@ private void checkLayoutRecursive(MemoryLayout layout) {
+                                inMessage(sl));
+                    }
+                    preceedingPadding = pl;
                 }
             }
-            checkGroupSize(sl, lastUnpaddedOffset);
+            checkNotAllPadding(sl);
+            checkGroup(sl, lastUnpaddedOffset);
         } else if (layout instanceof UnionLayout ul) {
             checkHasNaturalAlignment(layout);
             long maxUnpaddedLayout = 0;
             for (MemoryLayout member : ul.memberLayouts()) {
                 checkLayoutRecursive(member);
-                if (!(member instanceof PaddingLayout)) {
+                if (!(member instanceof PaddingLayout pl)) {
                     maxUnpaddedLayout = Long.max(maxUnpaddedLayout, member.byteSize());
+                } else {
+                    if (pl.byteSize() != ul.byteSize()) {
+                        throw new IllegalArgumentException("Superfluous padding " + pl + inMessage(ul));
+                    }
                 }
             }
-            checkGroupSize(ul, maxUnpaddedLayout);
+            checkGroup(ul, maxUnpaddedLayout);
         } else if (layout instanceof SequenceLayout sl) {
             checkHasNaturalAlignment(layout);
+            if (sl.elementLayout() instanceof PaddingLayout pl) {
+                throw memberException(sl, pl,
+                        "not supported because a sequence of a padding layout is not allowed");
+            }
             checkLayoutRecursive(sl.elementLayout());
         }
     }
 
-    // check for trailing padding
-    private void checkGroupSize(GroupLayout gl, long maxUnpaddedOffset) {
+    // check elements are not all padding layouts
+    private static void checkNotAllPadding(StructLayout sl) {
+        if (!sl.memberLayouts().isEmpty() && sl.memberLayouts().stream().allMatch(e -> e instanceof PaddingLayout)) {
+            throw new IllegalArgumentException("Layout '" + sl + "' is non-empty and only has padding layouts");
+        }
+    }
+
+    // check trailing padding
+    private static void checkGroup(GroupLayout gl, long maxUnpaddedOffset) {
         long expectedSize = Utils.alignUp(maxUnpaddedOffset, gl.byteAlignment());
         if (gl.byteSize() != expectedSize) {
             throw new IllegalArgumentException("Layout '" + gl + "' has unexpected size: "
                     + gl.byteSize() + " != " + expectedSize);
         }
     }
 
+    private static void assertIsAlignedBy(StructLayout gl, PaddingLayout padding, MemoryLayout element) {
+        if (padding.byteSize() > element.byteAlignment()) {
+            throw new IllegalArgumentException("The padding layout " + padding +
+                    " is not of minimum size to align " + element +
+                    "(with byte alignment " + element.byteAlignment() + ")" + inMessage(gl));
+        }
+    }
+
+    private static String inMessage(GroupLayout gl) {
+        return gl != null
+                ? " in " + gl
+                : "";
```

Comment:
Looks like this can be simpler now as well, since `gl` can never be `null` I think.